### PR TITLE
OS-36: Addition of API version and modules support

### PR DESCRIPTION
**Pull request by bartekkaminski**
The application architecture has been rebuilt as agreed

TASK:
https://openimis.atlassian.net/browse/OS-36

### CODE REVIEW

File: OpenImis.RestApi/Startup.cs [hunk 1]
```diff
@@ -32,10 +30,13 @@ public Startup(IConfiguration configuration)
 
         public void ConfigureServices(IServiceCollection services)
         {
-            // Add the DbContext 
-            //services.AddDbContext<IMISContext>(options => options.UseSqlServer(Configuration.GetConnectionString("IMISDatabase")));
+            var configImisModules = Configuration.GetSection("ImisModules").Get<List<ConfigImisModules>>();
+            int lastApiVersion = int.Parse(configImisModules[configImisModules.Count - 1].Version);
```
Comment (dragos-dobre) [inline OpenImis.RestApi/Startup.cs]: It doesn't means that the latest Module is configured at the end. We can add the new module configuration at the beginning.

File: OpenImis.RestApi/Startup.cs [hunk 2]
```diff
@@ -32,10 +30,13 @@ public Startup(IConfiguration configuration)
 
         public void ConfigureServices(IServiceCollection services)
         {
-            // Add the DbContext 
-            //services.AddDbContext<IMISContext>(options => options.UseSqlServer(Configuration.GetConnectionString("IMISDatabase")));
+            var configImisModules = Configuration.GetSection("ImisModules").Get<List<ConfigImisModules>>();
```
Comment (dragos-dobre) [inline OpenImis.RestApi/Startup.cs]: Should we sort the list by Version field?

File: OpenImis.RestApi/Security/IMISJwtSecurityTokenHandler.cs [hunk 3]
```diff
@@ -69,11 +77,15 @@ public ClaimsPrincipal ValidateToken(string securityToken, TokenValidationParame
             var tokenS = handler.ReadToken(securityToken) as JwtSecurityToken;
             int userId = Convert.ToInt32(tokenS.Claims.Where(w => w.Type == "UserId").Select(x => x.Value).FirstOrDefault());
 
-            UserData user = _imisModules.GetLoginModule().GetLoginLogic().GetById(userId);
+            string apiVersion = _httpContextAccessor.HttpContext.Request.Headers.Where(x => x.Key == "api-version").Select(s => s.Value).FirstOrDefault();
+
+            UserData user;
+            
+            if (apiVersion == null || apiVersion.StartsWith(_lastApiVersion.ToString())) user = (UserData)_imisModulesV2.GetLoginModule().GetLoginLogic().GetById(userId);
```
Comment (dragos-dobre) [inline OpenImis.RestApi/Security/IMISJwtSecurityTokenHandler.cs]: Why we need a variable _lastApiVersion if we know that this is for V2? 
What's happening when we have 3 or 4 versions of API (here we have only 2)?
Could we create the imisModules type based on a string here (i.e. "ModulesV"+_requiredApiVersion|_lastApiVersion+".IImisModules") and then make the call imisModules.GetLoginModule().GetLoginLogic().GetById(userId) using reflection?

File: OpenImis.RestApi/Startup.cs (hunk 1 above)
Comment (bartekkaminski) [reply OpenImis.RestApi/Startup.cs]: We can download the maximum value of the API in the file:
`var configImisModules = Configuration.GetSection("ImisModules").Get<List<ConfigImisModules>>();
int lastApiVersion = configImisModules.Max(c => int.Parse(c.Version));`

File: OpenImis.RestApi/Startup.cs (hunk 2 above)
Comment (bartekkaminski) [reply OpenImis.RestApi/Startup.cs]: This is not necessary here. The next line of code already get the version

File: OpenImis.RestApi/Security/IMISJwtSecurityTokenHandler.cs (hunk 3 above)
Comment (bartekkaminski) [reply OpenImis.RestApi/Security/IMISJwtSecurityTokenHandler.cs]: I would leave it as it is. When we have a finished version 2, we can finally come back and improve it. I will set the version 2 now rigidly

**Pull request merged**